src/utils/Subscription.js

```diff
@@ -1,45 +1,61 @@
 // encapsulates the subscription logic for connecting a component to the redux store, as
 // well as nesting subscriptions of descendant components, so that we can ensure the
 // ancestor components re-render before descendants
+
+function initListeners() {
+  let count = 0
+  let current = []
+  let next = []
+
+  return {
+    clear() {
+      count = 0
+      next = null
+      current = null
+    },
+
+    notify() {
+      current = next
+      for (let i = 0; i < count; i++) {
+        current[i]()
+      }
+    },
+
+    subscribe(listener) {
+      let isSubscribed = true
+      if (next === current) next = current.slice()
+      next.push(listener)
+      count++
+
+      return function unsubscribe() {
+        if (!isSubscribed || count === 0) return
```

Comment:
Counts always make me weary - especially when there's already an array that internally holds its count anyhow. Can we safely use `current.count` as the source of truth?